Automattic/woocommerce-payments
# WooPay compatibility with deferred intent creation UPE

Closes https://github.com/Automattic/woocommerce-payments/issues/6291

#### Changes proposed in this Pull Request
Currently, split UPE with the deferred intent creation on the server is implemented in all flows, including shortcode & Blocks checkout and _My account_ subpages, such as _Payment methods_, _Subscriptions_, _Pay for order_. The time has come for the WooPay to be compatible with this UPE type as well, thus this PR ensures that WooPay's functionality previously used only in pair with the legacy card gateway, is now entirely usable with the deferred intent creation UPE as well.









#### Testing instructions

> 💡 Ensure your WCPay Server listens to Stripe webhooks.

> Legacy card/legacy UPE/split UPE all disabled in WCPay Dev Tools. Only deferred intent creation UPE is enabled.
WooPay is enabled in Payments -> Settings. 
Only card payment method is enabled in Payments -> Settings. Act as a logged-in user. 

1️⃣ 

1. Add a product to the cart and navigate the shortcode checkout. 
2. On shortcode checkout, input default card details, billing details, and the email which isn’t yet registered with WooPay
3. Confirm that the _Save my information for a faster checkout_ checkbox is displayed under the card gateway. Don’t check this checkbox yet. 
4. Place the order and confirm that the order status is _Processing_, and on the Stripe side - payment in status _Succeeded_
5. Repeat steps 1-2, then check the  _Save my information for a faster checkout_ checkbox and place the order. 
6. Confirm you were redirected to an order placed page.
7. On WooPay's site admin panel, navigate to _Users_ and confirm that the user with the email you used in step 5 is now visible on the list. 
8. On the Stripe Customers page (e.g., `https://dashboard.stripe.com/test/customers`), confirm that the user is also listed.

Now, since you have the user registered in WooPay, check the following:

2️⃣  
1. Add a product to the cart and proceed to the shortcode checkout page again.
2. The email should be already provided but if not - provide the email of the previously registered user. 
3. Confirm that WooPay's pop-up modal appears on the page asking you to provide the OTP code for verification. 
4. After entering `000000` make sure you're redirected successfully to the checkout page with all the cart details.
5. After adding a new payment method/new shipping address if needed, on the WooPay checkout page, click _Place Order_ and confirm you're redirected to the merchant's order summary page stating the order was received successfully, and the payment method is a card with WooPay icon on the left. 
6. Navigate to `http://host.docker.internal:8090/account/orders` and confirm that the order placed is listed there.
7. Ensure the order status is correct in WC -> Orders and Stripe payments.
1. Log out after the previous steps, and navigate to the shortcode checkout so that the email field is empty
2. Click the WooPay Express Checkout button and enter the registered email
3. Complete the checkout with WooPay and confirm the order is successful

> Enable all the other available payment methods in _Payments -> Settings_, so that not only card appears on the list.

1. Perform the above two test blocks again with the list of enabled payment gateways. 

> Incognito window

Place an order with WooPay in the incognito window.

> Cart page

On the cart page, use the WooPay express checkout button and provide the email in the WooPay to checkout using WooPay. Confirm that the checkout went successfully.

> Product page

On the product page (e.g. a Beanie) use the express checkout button to test new user registration flow and checking out using the existing email address.

> Blocks checkout

Perform testing instructions from 1️⃣ and 2️⃣ on the Blocks checkout page.

> Other UPEs

* Enable split UPE and disable deferred intent creation UPE. Enable WooPay. Go to the checkout page and confirm that the legacy card gateway is loaded as expected. Disable WooPay and confirm that the card gateway (Stripe Payment Element) is rendered. 
* Perform sanity checks for the split UPE (e.g. buy a product and a subscription to double-check there's no regression)
* Perform sanity checks for the legacy UPE (same as above)


---- 
`class-checkout-service.php` was extended a bit. Below are the changes and how to test them (and changes in `class-payment-information` as well):
1. `create_intention_request` method is used for all the legacy card payments and card payments within the deferred intent creation. This means, that you will probably have this method tested after all the above steps but if still unsure - just enable legacy gateway (non-UPE) or deferred intent UPE and purchase any product with a card.
2. `create_and_confirm_setup_intention_request` was slightly modified as well. To test it out, enable legacy gateway (non-UPE) or deferred intent creation UPE, and purchase a free trial subscription. Ensure that the payment was processed successfully.





-------------------


**Post merge**
